File: house-robber/house-robber.cpp
class Solution {
public:
    int solve(vector<int>& nums, int idx, vector<int>& dp) {
        if (idx<0)
            return 0;
        if (dp[idx]>=0)
            return dp[idx];
        // either we can steal from 2 homes before+this house.
        // or skip this house and steal from 1 house before. 
        // that gives us the most optimized answer till that house.n 
        dp[idx] = max(solve(nums, idx-2, dp)+nums[idx], solve(nums, idx-1, dp));
        return dp[idx];
    }
    
    int rob(vector<int>& nums) {
        vector<int> dp(nums.size(), -1);
        return solve(nums, nums.size()-1, dp);
    }
};

File: house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums: return 0
        
        # robbing house at ith index gives us 0
        # question is, which other alternate houses to rob, and which particular permutation sequence gives us max profit?
        # we use dp to solve the same
        
        if len(nums)<=2: return max(nums)
        
        dp = [0]*len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[:2])
        
        for i in range(2, len(nums)):
            dp[i] = max(dp[i-1], nums[i]+dp[i-2])
            
        return dp[-1]